x/poolmanager/taker_fee.go

```diff
@@ -0,0 +1,182 @@
+
+	appparams "github.com/osmosis-labs/osmosis/v19/app/params"
+
+	"github.com/osmosis-labs/osmosis/osmoutils"
+	"github.com/osmosis-labs/osmosis/v19/x/poolmanager/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v19/x/txfees/types"
+)
+
+// SetDenomPairTakerFee sets the taker fee for the given trading pair.
+// If the taker fee for this denom pair matches the default taker fee, then
+// it is deleted from state.
+func (k Keeper) SetDenomPairTakerFee(ctx sdk.Context, denom0, denom1 string, takerFee sdk.Dec) {
+	store := ctx.KVStore(k.storeKey)
+	if takerFee.Equal(k.GetParams(ctx).TakerFeeParams.DefaultTakerFee) {
+		store.Delete(types.FormatDenomTradePairKey(denom0, denom1))
+		return
+	} else {
+		osmoutils.MustSetDec(store, types.FormatDenomTradePairKey(denom0, denom1), takerFee)
+	}
+}
+
+// SenderValidationSetDenomPairTakerFee sets the taker fee for the given trading pair iff the sender's address
+// also exists in the pool manager taker fee admin address list.
+func (k Keeper) SenderValidationSetDenomPairTakerFee(ctx sdk.Context, sender, denom0, denom1 string, takerFee sdk.Dec) error {
+	adminAddresses := k.GetParams(ctx).TakerFeeParams.AdminAddresses
+	isAdmin := false
+	for _, admin := range adminAddresses {
+		if admin == sender {
+			isAdmin = true
+			break
+		}
+	}
+	if !isAdmin {
+		return fmt.Errorf("%s is not in the pool manager taker fee admin address list", sender)
+	}
+
+	k.SetDenomPairTakerFee(ctx, denom0, denom1, takerFee)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeMsgSetDenomPairTakerFee,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, sender),
+			sdk.NewAttribute(types.AttributeKeyDenom0, denom0),
+			sdk.NewAttribute(types.AttributeKeyDenom1, denom1),
+			sdk.NewAttribute(types.AttributeKeyTakerFee, takerFee.String()),
+		),
+	})
+
+	return nil
+}
+
+// GetTradingPairTakerFee returns the taker fee for the given trading pair.
+// If the trading pair does not exist, it returns the default taker fee.
+func (k Keeper) GetTradingPairTakerFee(ctx sdk.Context, denom0, denom1 string) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FormatDenomTradePairKey(denom0, denom1)
+
+	takerFee := &sdk.DecProto{}
+	found, err := osmoutils.Get(store, key, takerFee)
+	if err != nil {
+		return sdk.Dec{}, err
+	}
+	if !found {
+		return k.GetParams(ctx).TakerFeeParams.DefaultTakerFee, nil
+	}
+
+	return takerFee.Dec, nil
+}
+
+// extractTakerFeeAndDistribute takes in a pool and extracts the taker fee from the pool and sends it to the non native fee pool module account.
+// Its important to note here that in the original swap, the taker fee + spread fee is sent to the pool's address, so this is why we
+// pull directly from the pool and not the user's account.
+func (k Keeper) extractTakerFeeAndDistribute(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, sender sdk.AccAddress, exactIn bool) (sdk.Coin, error) {
```

Comment:
IMO taker fee extraction and distribution are implementation details that should not be reflected in the API. Charging abstracts this more generally

```suggestion
func (k Keeper) chargeTakerFee(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, sender sdk.AccAddress, exactIn bool) (sdk.Coin, error) {
```